crates/edr_napi/src/solidity_tests.rs

```diff
@@ -0,0 +1,111 @@
+mod runner;
+mod test_results;
+mod test_suite;
+
+use std::path::Path;
+
+use forge::TestFilter;
+use napi::{
+    threadsafe_function::{
+        ErrorStrategy, ThreadSafeCallContext, ThreadsafeFunction, ThreadsafeFunctionCallMode,
+    },
+    tokio,
+    tokio::runtime,
+    Env, JsFunction,
+};
+use napi_derive::napi;
+
+use crate::solidity_tests::{
+    runner::build_runner, test_results::SuiteResult, test_suite::TestSuite,
+};
+
+/// Executes solidity tests.
+#[napi]
+pub struct SolidityTestRunner {
+    /// The callback to call with the results as they become available.
+    results_callback_fn: ThreadsafeFunction<SuiteResult>,
+}
+
+// The callback has to be passed in the constructor because it's not `Send`.
+#[napi]
+impl SolidityTestRunner {
+    #[doc = "Creates a new instance of the SolidityTestRunner. The callback function will be called with suite results as they finish."]
+    #[napi(constructor)]
+    pub fn new(env: Env, results_callback: JsFunction) -> napi::Result<Self> {
+        let mut results_callback_fn: ThreadsafeFunction<_, ErrorStrategy::CalleeHandled> =
+            results_callback.create_threadsafe_function(
+                // Unbounded queue size
+                0,
+                |ctx: ThreadSafeCallContext<SuiteResult>| Ok(vec![ctx.value]),
+            )?;
+
+        // Allow the event loop to exit before the function is destroyed
+        results_callback_fn.unref(&env)?;
+
+        Ok(Self {
+            results_callback_fn,
+        })
+    }
+
+    #[doc = "Runs the given test suites."]
+    #[napi]
+    pub async fn run_tests(&self, test_suites: Vec<TestSuite>) -> napi::Result<Vec<SuiteResult>> {
+        let test_suites = test_suites
+            .into_iter()
+            .map(|item| Ok((item.id.try_into()?, item.contract.try_into()?)))
+            .collect::<Result<Vec<_>, napi::Error>>()?;
+        let mut runner = build_runner(test_suites)?;
+
+        let (tx_results, mut rx_results) =
+            tokio::sync::mpsc::unbounded_channel::<(String, forge::result::SuiteResult)>();
+        let (tx_end_result, mut rx_end_result) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
Question: I raised this in chat as well, but should this be an unbounded channel or a oneshot?

Reply:
Yeah thanks, I've changed that locally and I'll push one more commit to fix the progress reporting only happening at the end. Will reping for review